aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -0,0 +1,98 @@
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs?highlight=10)]
+
+> [!NOTE]
+> The example above specifies a default model name. This can be omitted if you always specify the `Name` on the `ModelBinderAttribute`.
+
+With this in place, action methods can use the `Author` entity directly:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/BoundAuthorsController.cs?highlight=2)]
+
+Using this approach, both the controller and action method are simplified compared to performing the same work in the controller:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/AuthorsController.cs?highlight=2)]
+
+Instead of applying an attribute, you can implement [IModelBinderProvider](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.modelbinding.imodelbinderprovider). This is how the built-in framework binders are implemented. Note that when you specify the type your binder operates on, you specify the type of argument it will produce, not the input your binder will accept. The following binder provider would work for the `AuthorEntityBinder` implemented above, and when added to MVC's collection of providers, would eliminate the need for the `ModelBinderAttribute` on `Author`.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinderProvider.cs?highlight=17-20)]
+
+> [!NOTE]
+> The example above returns a `BinderTypeModelBinder`, which acts as a factory for model binders and provides dependency injection for them. Use it if your model binder requires services from DI.
+
+In order to configure MVC to use a custom model binder provider, you add it in `ConfigureServices`. When evaluating model binders, the collection of providers is examined in order. The first provider that returns a binder will be used. Below you can see the default model binders, in order:
+
+![default model binders](custom-model-binding/images/default-model-binders.png "default model binders")
+
+
+Adding your provider to the end of the collection may result in it never being called. In this example, the custom provider is added to the beginning of the collection to ensure it is used for `Author` action arguments.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Startup.cs?name=callout&highlight=5-9)]
+
+## Recommendations and best practices
+
+Model binders should not attempt to set status codes or return results (for example, 404 Not Found). Instead, if model binding fails, this should be the responsibility of an [action filter](/mvc/controllers/filters.md) or logic within the action method itself.
```

Comment:
~Instead~
If model bind fails, the [action filter](/mvc/controllers/filters.md) or logic within the action method should handle the failure.